sql: use wildcard to show grants for all schemas in a database

lkyuchukov opened the pull request:
Fixes [#80253](https://github.com/cockroachdb/cockroach/issues/80253)

Release note: 
Allow the use wildcard to show grants for all schemas in a database.
```
> show grants on schema foo.*;
  database_name | schema_name | grantee | privilege_type | is_grantable
----------------+-------------+---------+----------------+---------------
  foo           | bar         | admin   | ALL            |     true
  foo           | bar         | root    | ALL            |     true
  foo           | baz         | admin   | ALL            |     true
  foo           | baz         | root    | ALL            |     true
  foo           | public      | admin   | ALL            |     true
  foo           | public      | public  | CREATE         |    false
  foo           | public      | public  | USAGE          |    false
  foo           | public      | root    | ALL            |     true
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/80861)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Hi @lkyuchukov,
Thanks for your contribution!

Please make sure that you have a single commit. Instead of picking up changes to master with `git merge` you'll want to use `git rebase`. Take a look at this wiki page for more details: https://cockroachlabs.atlassian.net/l/c/E0mzbdeQ.

Please also add some tests, probably in `grant_schema` (https://github.com/cockroachdb/cockroach/blob/b266fab64726c31d53bb837818f60e81d502ba89/pkg/sql/logictest/testdata/logic_test/grant_schema)

Otherwise, the changes look good to me, modulo a few nits in comments below. I think @cockroachdb/sql-schema and @cockroachdb/sql-experience will also want to review this change.

> Reviewed 9 of 9 files at r1, 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-:-N1FVJr9BOG_X6Cb4JZS:b-jmcb2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fabiog1901 and @lkyuchukov)

___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1FZl-6E9ISsVQ6zRRa):*
This commit should be squashed with the previous.

Commit title should be `package_name: short description of change`. Your PR title works well (`sql: use wildcard to show grants for all schemas in a database`).

Please add a release note `Release note (sql change): ...` with an explanation of the user-facing change. The category (sql change) is important for our documentation team.

You can update the PR description to match the commit message.

Please see this wiki page for more details: https://cockroachlabs.atlassian.net/l/c/Sve6VN7Z
___
*[`pkg/sql/opt_catalog.go` line 157 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1FWtL65yRkBcWDguSq:-N1FWtL65yRkBcWDguSr:b-t8735y) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c24ee2a9bb0933eafda9749a1b44e16c3977197/pkg/sql/opt_catalog.go#L157)):*
> ```Go
> }
> 
> func (oc *optCatalog) GetAllSchemaNamesForDB(ctx context.Context, dbName string) ([]cat.SchemaName, error) {
> ```

nit: add a comment
```
// GetAllSchemaNamesForDB is part of the cat.Catalog interface.
```

nit: move this function so that its order in the file matches the order in the interface (i.e. below `ResolveSchema`)
___
*[`pkg/sql/resolver.go` line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1FWiLcEYblMSFI9phm:-N1FWiLcEYblMSFI9phn:bmon11q) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c24ee2a9bb0933eafda9749a1b44e16c3977197/pkg/sql/resolver.go#L161)):*
> ```Go
> }
> 
> func (p *planner) GetSchemasForDB(ctx context.Context, dbName string, ) (map[descpb.ID]string, error) {
> ```

nit: add a comment
```
// GetSchemasForDB ...
```
___
*[`pkg/sql/opt/cat/catalog.go` line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1FXDFf16hdNJcf5y5o:-N1FXDFf16hdNJcf5y5p:b6xuf9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c24ee2a9bb0933eafda9749a1b44e16c3977197/pkg/sql/opt/cat/catalog.go#L88)):*
> ```Go
> 	ResolveSchema(ctx context.Context, flags Flags, name *SchemaName) (Schema, SchemaName, error)
> 
> 	// GetAllSchemaNamesForDB Get all the SchemaNames for a database
> ```

nit: Get -> gets
nit: comments should end with a period.
___
*[`pkg/sql/opt/testutils/testcat/test_catalog.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1FXV9F1b9Jru3_0nVZ:-N1FXV9F1b9Jru3_0nV_:b-tq1xs2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c24ee2a9bb0933eafda9749a1b44e16c3977197/pkg/sql/opt/testutils/testcat/test_catalog.go#L53)):*
> ```Go
> }
> 
> func (tc *Catalog) GetAllSchemaNamesForDB(ctx context.Context, dbName string) ([]cat.SchemaName, error) {
> ```

nit: move this function so that its order in the file matches the order in the interface (i.e. below `ResolveSchema`)
___
*[`pkg/sql/parser/sql.y` line 6868 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1FYpH3C5AlWblu6Wxx:-N1FYpH3C5AlWblu6Wxy:b-vqgz40) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c24ee2a9bb0933eafda9749a1b44e16c3977197/pkg/sql/parser/sql.y#L6868)):*
> ```Yacc
>   }
> | SCHEMA schema_wildcard
> 	{
> ```

nit: remove tabs
___
*[`pkg/sql/parser/types_regex.tmp` line 0 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1FZeO189iSSQ8Ir5sH:-N1FZeO189iSSQ8Ir5sI:b-3ugar3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c7704061ff4e54e1ab0e7a73f754b52cdfa955c/pkg/sql/parser/types_regex.tmp#L0)):*
Did you mean to add this file?


<!-- Sent from Reviewable.io -->

lkyuchukov committed:
sql: use wildcard to show grants for all schemas in a database

Release note(sql change):
Allow the use wildcard to show grants for all schemas in a database.

rytaft reviewed (commented):
Thanks for making updates @lkyuchukov! Just a couple more nits below.

I still want someone from @cockroachdb/sql-schema and/or @cockroachdb/sql-experience to take a look at this. cc @ajwerner @rafiss 

> Reviewed 1 of 7 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-:-N1Jnwd34i1gbUq9G4Sz:b6pqetl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fabiog1901, @lkyuchukov, and @rytaft)

___
*[`-- commits` line 4 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1G231DC3piKAadh2dl):*
This should be `Release note (sql change):`  
The category (sql change) is important for our documentation team.
___
*[`pkg/sql/logictest/testdata/logic_test/grant_schema` line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1JnINp-FSn56SqRH4P:-N1JnINp-FSn56SqRH4Q:bx93kcp) ([raw file](https://github.com/cockroachdb/cockroach/blob/56de04f750fd3f720cd35fc6d1887d61d83b9c24/pkg/sql/logictest/testdata/logic_test/grant_schema#L35)):*
> ```
> 
> query TTTTB colnames
> SHOW GRANTS ON SCHEMA test.*
> ```

would be good to also add a test for databases with multiple schemas (e.g. create a user-defined schema and make sure it shows up)


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80861)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fabiog1901, @lkyuchukov, and @rytaft)

___
*[`-- commits` line 4 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1LLMny8-XisC6xs1kc):*
Looks like you lost a couple of characters here.  Please update to say `Release note (sql change):`
___
*[`pkg/sql/resolver.go` line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-N1FWiLcEYblMSFI9phm:-N1LLutS-oEBdHvil_jr:bhv413b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c24ee2a9bb0933eafda9749a1b44e16c3977197/pkg/sql/resolver.go#L161)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: add a comment
```
// GetSchemasForDB ...
```

</blockquote></details>

Still missing a comment here


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-:-N1PAn4zCNw24YRaCCLF:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @fabiog1901 and @rytaft)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
this looks good to me! thanks for your contribution

bors r+

> Reviewed 3 of 9 files at r1, 5 of 6 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80861#-:-N1QhwxW40ojPVEKy4t8:b-728gfa)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @fabiog1901, @postamar, and @rytaft)



<!-- Sent from Reviewable.io -->
